include/boost/random/niederreiter_base2.hpp

```diff
@@ -0,0 +1,391 @@
+/* boost random/nierderreiter_base2.hpp header file
+ *
+ * Copyright Justinas Vygintas Daugmaudis 2010-2017
+ * Distributed under the Boost Software License, Version 1.0. (See
+ * accompanying file LICENSE_1_0.txt or copy at
+ * http://www.boost.org/LICENSE_1_0.txt)
+ */
+
+#ifndef BOOST_RANDOM_NIEDERREITER_BASE2_HPP
+#define BOOST_RANDOM_NIEDERREITER_BASE2_HPP
+
+#include <boost/random/detail/gray_coded_qrng_base.hpp>
+#include <boost/random/detail/config.hpp>
+#include <boost/random/detail/operators.hpp>
+
+#include <limits>
+#include <boost/cstdint.hpp>
+
+#include <boost/dynamic_bitset.hpp>
+#include <boost/multi_array.hpp>
+
+//!\file
+//!Describes the quasi-random number generator class template niederreiter_base2.
+//!
+//!\b Note: it is especially useful in conjunction with class template uniform_real.
+
+namespace boost {
+namespace random {
+
+/** @cond */
+namespace detail {
+namespace nb2 {
+
+/*
+  Primitive polynomials in binary encoding
+  {
+    { 1, 0, 0, 0, 0, 0 },    1
+    { 0, 1, 0, 0, 0, 0 },    x
+    { 1, 1, 0, 0, 0, 0 },    1 + x
+    { 1, 1, 1, 0, 0, 0 },    1 + x + x^2
+    { 1, 1, 0, 1, 0, 0 },    1 + x + x^3
+    { 1, 0, 1, 1, 0, 0 },    1 + x^2 + x^3
+    { 1, 1, 0, 0, 1, 0 },    1 + x + x^4
+    { 1, 0, 0, 1, 1, 0 },    1 + x^3 + x^4
+    { 1, 1, 1, 1, 1, 0 },    1 + x + x^2 + x^3 + x^4
+    { 1, 0, 1, 0, 0, 1 },    1 + x^2 + x^5
+    { 1, 0, 0, 1, 0, 1 },    1 + x^3 + x^5
+    { 1, 1, 1, 1, 0, 1 },    1 + x + x^2 + x^3 + x^5
+    { 1, 1, 1, 0, 1, 1 }     1 + x + x^2 + x^4 + x^5
+  };
+*/
+
+// Maximum allowed space dimension
+#define BOOST_RANDOM_NIEDERREITER_BASE2_MAX_DIMENSION 54
+
+struct niederreiter_tables
+{
+  BOOST_STATIC_CONSTANT(int, max_dimension = BOOST_RANDOM_NIEDERREITER_BASE2_MAX_DIMENSION);
+
+  // Binary irreducible polynomials (primes in the ring GF(2)[X]), evaluated at X=2. 
+  static unsigned short polynomial(std::size_t n)
+  {
+    static const unsigned short nb2_a[max_dimension] = {
+      2, 3, 7, 11, 13, 19, 25, 31, 37, 41,
+      47, 55, 59, 61, 67, 73, 87, 91, 97, 103,
+      109, 115, 117, 131, 137, 143, 145, 157,
+      167, 171, 185, 191, 193, 203, 211, 213,
+      229, 239, 241, 247, 253, 283, 285, 299, 
+      301, 313, 319, 333, 351, 355, 357, 361, 369, 
+      375
+    };
+
+    return nb2_a[n];
+  }
+};
+
+// Return the base 2 logarithm for a given bitset v
```

Comment:
Can `boost/integer/integer_log2.hpp` be used here instead? Maybe not because you are using a bitset?

Reply:
Yes, precisely. dynamic bitset log here. This implementation is quite good if I may say so, but I wouldn't mind if it was moved to some other header.